### PR TITLE
[WIP] Improve Code Coverage to 99% or 100%

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

Work in progress PR for #1335 
Contributions welcome! I'm also happy to mentor anyone looking to write some tests of their own.

### CODE REVIEW

File: test/route.test.js [hunk 1]
```diff
@@ -54,6 +54,19 @@ test('route', t => {
     }
   })
 
+  test('invalid handler attribute - route', t => {
+    // Unique test case where a valid handler function is passed in
+    // the handler argument, but the options object contains a handler
+    // property that is not a function. Will throw an error on line 537 in fastify.js
```
Comment (allevo) [inline test/route.test.js]: Could you remove this comment?
The test is readable as is 😄

File: test/emit-warning.test.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    t.strictEqual(string, `Using 'genReqId' in logger options is deprecated. Use fastify options instead. See: https://www.fastify.io/docs/latest/Server/#gen-request-id`)
+  }
+  Fastify({ logger: { genReqId: 'test' } })
+})
+
+test('should emit warning if basePath prop is used', t => {
+  t.plan(1)
+  process.emitWarning = function (string) {
+    t.strictEqual(string, `basePath is deprecated. Use prefix instead. See: https://www.fastify.io/docs/latest/Server/#prefix`)
+  }
```
Comment (mcollina) [inline test/emit-warning.test.js]: We should not monkey-patch emitWarning, as it can collide with multiple tests.
You should add a listener to process.once(‘warning’) instead, and remove such a listener on a tearDown

**Comment by Ethan-Arrowood** [comment]
Any ideas on how to test this chunk of code? https://github.com/fastify/fastify/blob/81a5551763d14f7424b234747909bf6dcc1340a7/fastify.js#L873-L877

Been struggling with this one for a while and not sure how to make this logic branch execute

**Comment by mcollina** [comment]
So, you need to specify a route using a `preHandler`, like so:

```js
fastify.get('/', {
  preHandler (req, reply, cb) {
    cb()
  }
}
```

And check with `process.once('warning')` that a warning is emitted.

**Comment by Ethan-Arrowood** [comment]
Woo! fastify.js is 99.2% now with my latest commit. I also added a section to testing docs about running and debugging fastify cause this is something I think might be useful for developers to use in the future.

**Comment by Ethan-Arrowood** [comment]
Any ideas on how to test this line? 
https://github.com/fastify/fastify/blob/adfa0c32b818af0f6f477dce5680bc871a5e2669/lib/handleRequest.js#L9

Coverage reports its being hit 737 times but never evaluating true. 

Climbing through the call stack it looks like the line:
https://github.com/fastify/fastify/blob/adfa0c32b818af0f6f477dce5680bc871a5e2669/fastify.js#L426

Is evaluating true 751 times and would be preventing the `handleRequest` line from being reached.

**Comment by Ethan-Arrowood** [comment]
Are these lines reachable?
https://github.com/fastify/fastify/blob/adfa0c32b818af0f6f477dce5680bc871a5e2669/lib/validation.js#L41
https://github.com/fastify/fastify/blob/adfa0c32b818af0f6f477dce5680bc871a5e2669/lib/validation.js#L50

**Comment by mcollina** [comment]
You can refactor those lines to use Object.keys(). It’s possible to trigger those lines, but I don’t think it’s useful at all.

File: lib/reply.js [hunk 3]
```diff
@@ -80,7 +80,8 @@ Object.defineProperty(Reply.prototype, 'sent', {
 
     this[kReplySentOverwritten] = true
     this[kReplySent] = true
-  }
+  },
+  configurable: true
```
Comment (jsumners) [inline lib/reply.js]: Help me understand. This property has both a getter and a setter. That should negate the need for `configurable: true`. Why are you setting this option?

Comment (jsumners) [reply lib/reply.js]: That's not correct. The `configurable` property dictates if the property can be deleted. Run this:

```js
'use strict'

const foo = {}
Object.defineProperty(foo, 'bar', {
  enumerable: true,
  get () {
    return this[Symbol.for('bar')]
  },
  set(value) {
    console.log('setting bar')
    this[Symbol.for('bar')] = value
  }
})

console.log(`foo = ${JSON.stringify(foo)}`)
foo.bar = 'baz'
console.log(`foo = ${JSON.stringify(foo)}`)
```

Notice that it works as expected. Now run this:

```js
'use strict'

function Foo () {}

Object.defineProperty(Foo.prototype, 'bar', {
  enumerable: true,
  get() {
    return this[Symbol.for('bar')]
  },
  set(value) {
    console.log('setting bar')
    this[Symbol.for('bar')] = value
  }
})

const foo = new Foo()
console.log(`foo = ${JSON.stringify(foo)}`)
foo.bar = 'baz'
console.log(`foo = ${JSON.stringify(foo)}`)
console.log(`foo.bar = ${foo.bar}`)
```

Notice that both of the stringify lines output `{}`, but the last one outputs `foo.bar = baz`. The property _is_ being modified. It is just being modified on the prototype, not the instance.

Comment (jsumners) [reply lib/reply.js]: For more fun with properties, study https://gist.github.com/jsumners/0a586bf6c3dcb835d692251347a0e730

File: lib/validation.js [hunk 4]
```diff
@@ -37,19 +37,14 @@ function build (context, compile, schemas) {
     // The header keys are case insensitive
     //  https://tools.ietf.org/html/rfc2616#section-4.2
     const headersSchemaLowerCase = {}
-    for (const k in headers) {
-      if (headers.hasOwnProperty(k) !== true) continue
-      headersSchemaLowerCase[k] = headers[k]
-    }
+    Object.keys(headers).forEach(k => headersSchemaLowerCase[k] = headers[k])
```
Comment (jsumners) [inline lib/validation.js]: Increasing test coverage should _not_ include changing the code being tested. This sort of thing should be addressed in another PR.

Comment (Ethan-Arrowood) [reply lib/validation.js]: Fair point! Would you like me to revert this change and make it in a separate PR?

Comment (jsumners) [reply lib/validation.js]: That would be most appropriate.

Comment (Ethan-Arrowood) [reply lib/validation.js]: Okay removed commit; going to add it to another PR now

File: lib/reply.js (hunk 3 above)
Comment (Ethan-Arrowood) [reply lib/reply.js]: 😣 oh javascript

File: docs/Testing.md [hunk 5]
```diff
@@ -148,3 +148,19 @@ tap.test('GET `/` route', async (t) => {
   t.deepEqual(response.body, { hello: 'world' })
 })
 ```
+
+### How to inspect `fastify.js` while testing
```
Comment (mcollina) [inline docs/Testing.md]: I think this should be named "How to inspect tap tests" instead.
This has nothing related to fastify, but I understand why it's good to have this info here.

File: test/inject.test.js [hunk 6]
```diff
@@ -391,3 +391,24 @@ test('should error the promise if ready errors', t => {
...
+  t.plan(2)
+  const fastify = Fastify()
+  const error = new Error('kaboom')
+
+  fastify.register((instance, opts) => {
+    return Promise.reject(error)
+  })
+  // }).after(function () {
+  //   t.pass('after is called')
+  // })
```
Comment (mcollina) [inline test/inject.test.js]: why is this commented out?

File: lib/reply.js (hunk 3 above)
Comment (mcollina) [reply lib/reply.js]: If the function is run but the report does not show it up, then let's just flag it as ignored by istanbul.

File: test/inject.test.js (hunk 6 above)
Comment (Ethan-Arrowood) [reply test/inject.test.js]: Oh hmm, my mistake - will fix in next commit

**Comment by mcollina** [comment]
Can you rebase, there is a conflict.

How are we in regards to code coverage with these changes?

**Comment by Ethan-Arrowood** [comment]
screen shot 2019-01-08 at 11 43 35 am

**Comment by Ethan-Arrowood** [comment]
Sorry! Haven’t had access to my dev laptop for a day now and i don’t know when I’ll have it back (it’s getting repaired). Yes we can close and merge this. I don’t know if we are at 99% yet but I think it’ll be good to finally merge this in and then open separate PR’s for any additional coverage increases (wrapThenable.js should be an easy file to test but doesn’t have much coverage at the moment)
